<a href="https://colab.research.google.com/github/manasamorthad/Machine-Learning/blob/main/ensemble.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [22]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [23]:
df=pd.read_csv("/content/diabetes (2).csv")

In [24]:
[col for col in df.columns if df[col].isnull().sum()>0]

[]

In [7]:
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [25]:
X=df.drop("Outcome",axis=1)
y=df["Outcome"]

In [26]:
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=42)

In [27]:
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

In [38]:
clf=DecisionTreeClassifier(random_state=42)
clf=clf.fit(x_train,y_train)
y_pred=clf.predict(x_test)

In [39]:
print("accuracy is",metrics.accuracy_score(y_test,y_pred))
print("precison is",metrics.precision_score(y_test,y_pred))
print("recall is",metrics.recall_score(y_test,y_pred))
print("f1-score is",metrics.f1_score(y_test,y_pred))

accuracy is 0.7012987012987013
precison is 0.5555555555555556
recall is 0.6875
f1-score is 0.6145251396648045


In [40]:
from sklearn.ensemble import RandomForestClassifier

In [66]:
rf=RandomForestClassifier(n_estimators=50,random_state=0)
rf.fit(x_train,y_train)
y_pred_rf=rf.predict(x_test)

In [67]:
print("accuracy is",metrics.accuracy_score(y_test,y_pred_rf))
print("precison is",metrics.precision_score(y_test,y_pred_rf))
print("recall is",metrics.recall_score(y_test,y_pred_rf))
print("f1-score is",metrics.f1_score(y_test,y_pred_rf))

accuracy is 0.7316017316017316
precison is 0.618421052631579
recall is 0.5875
f1-score is 0.6025641025641025


| Metric        | Decision Tree | Random Forest |
| ------------- | ------------- | ------------- |
| **Accuracy**  | 0.7013        | 0.7316        |
| **Precision** | 0.5556        | 0.6184        |
| **Recall**    | 0.6875        | 0.5875        |
| **F1-score**  | 0.6145        | 0.6026        |
        |


In [47]:
rf=RandomForestClassifier(  n_estimators=200,
    max_depth=10,
    min_samples_split=5,
    min_samples_leaf=2,
    max_features='sqrt',
    criterion='entropy',
    random_state=0)
rf.fit(x_train,y_train)
y_pred_rf=rf.predict(x_test)

In [48]:
print("accuracy is",metrics.accuracy_score(y_test,y_pred_rf))
print("precison is",metrics.precision_score(y_test,y_pred_rf))
print("recall is",metrics.recall_score(y_test,y_pred_rf))
print("f1-score is",metrics.f1_score(y_test,y_pred_rf))

accuracy is 0.7619047619047619
precison is 0.654320987654321
recall is 0.6625
f1-score is 0.6583850931677019


| **Metric**    | **DT**     | **RF (50)** | **RF (Tuned)** |
| ------------- | ---------- | ----------- | -------------- |
| **Accuracy**  | 0.7013     | 0.7316      | **0.7619**     |
| **Precision** | 0.5556     | 0.6184      | **0.6543**     |
| **Recall**    | **0.6875** | 0.5875      | 0.6625         |
| **F1-Score**  | 0.6145     | 0.6026      | **0.6584**     |


In [68]:
from sklearn.linear_model import LogisticRegression

In [71]:
from sklearn.ensemble import VotingClassifier

In [73]:
estimators = [
    ('lr', LogisticRegression(max_iter=200, random_state=42)),
    ('dt', DecisionTreeClassifier(random_state=42)),
    ('rf', RandomForestClassifier(n_estimators=100, random_state=42))
]

# Create Voting Classifier - Hard Voting
hard_voting = VotingClassifier(estimators=estimators, voting='hard')

# Train the ensemble
hard_voting.fit(x_train, y_train)

# Make predictions
y_pred = hard_voting.predict(x_test)

# Evaluate Accuracy
score = metrics.accuracy_score(y_test, y_pred)
print("Hard Voting Accuracy: {:.3f}".format(score))


Hard Voting Accuracy: 0.749


In [75]:
soft_voting = VotingClassifier(estimators=estimators, voting='soft')
soft_voting.fit(x_train, y_train)
y_pred_soft = soft_voting.predict(x_test)
print("Soft Voting Accuracy: {:.3f}".format(metrics.accuracy_score(y_test, y_pred_soft)))


Soft Voting Accuracy: 0.736


In [77]:
lr = LogisticRegression(max_iter=200, random_state=42)
dt = DecisionTreeClassifier(random_state=42)
rf = RandomForestClassifier(n_estimators=100, random_state=42)

In [78]:
lr.fit(x_train, y_train)
dt.fit(x_train, y_train)
rf.fit(x_train, y_train)

w1 = metrics.accuracy_score(y_test, lr.predict(x_test))
w2 = metrics.accuracy_score(y_test, dt.predict(x_test))
w3 = metrics.accuracy_score(y_test, rf.predict(x_test))

weighted_voting = VotingClassifier(
    estimators=estimators,
    voting='soft',           # soft voting with probabilities
    weights=[w1, w2, w3]     # weights based on model performance
)
weighted_voting.fit(x_train, y_train)
y_pred_weighted = weighted_voting.predict(x_test)
print("Weighted Voting Accuracy: {:.3f}".format(metrics.accuracy_score(y_test, y_pred_weighted)))

Weighted Voting Accuracy: 0.736


In [80]:
from sklearn.neighbors import KNeighborsClassifier

In [81]:
estimators = [
    ('lr', LogisticRegression(max_iter=200, random_state=42)),
    ('dt', DecisionTreeClassifier(random_state=42)),
    ('knn', KNeighborsClassifier(n_neighbors=5))
]

# --- Hard Voting ---
hard_voting = VotingClassifier(estimators=estimators, voting='hard')
hard_voting.fit(x_train, y_train)
y_pred_hard = hard_voting.predict(x_test)

# --- Soft Voting ---
soft_voting = VotingClassifier(estimators=estimators, voting='soft')
soft_voting.fit(x_train, y_train)
y_pred_soft = soft_voting.predict(x_test)

# --- Evaluation ---
print("Hard Voting Metrics:")
print("Accuracy:", round(metrics.accuracy_score(y_test, y_pred_hard),3))
print("Precision:", round(metrics.precision_score(y_test, y_pred_hard, average='macro'),3))
print("Recall:", round(metrics.recall_score(y_test, y_pred_hard, average='macro'),3))
print("F1-Score:", round(metrics.f1_score(y_test, y_pred_hard, average='macro'),3))
print()

print("Soft Voting Metrics:")
print("Accuracy:", round(metrics.accuracy_score(y_test, y_pred_soft),3))
print("Precision:", round(metrics.precision_score(y_test, y_pred_soft, average='macro'),3))
print("Recall:", round(metrics.recall_score(y_test, y_pred_soft, average='macro'),3))
print("F1-Score:", round(metrics.f1_score(y_test, y_pred_soft, average='macro'),3))

Hard Voting Metrics:
Accuracy: 0.719
Precision: 0.692
Recall: 0.7
F1-Score: 0.695

Soft Voting Metrics:
Accuracy: 0.71
Precision: 0.686
Recall: 0.696
F1-Score: 0.689
